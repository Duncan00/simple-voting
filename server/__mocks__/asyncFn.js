'use strict';

const util = require('util');

const asyncFn = fname =>
	jest.fn(async (...args) => {
		const msg = `unexpected call to ${fname} with (${util.inspect(
			{args},
			10
		)}), did you miss any mockImplementationOnce?`;
		throw new Error(msg);
	});

module.exports = asyncFn;
