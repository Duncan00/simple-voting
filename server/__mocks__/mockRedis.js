'use strict';

const Redis = require('ioredis-mock');
const redis = new Redis();
// const asyncFn = require('./asyncFn');
//
// const redis = {
// 	mock_object_name: 'mock-redis-object',
// 	incr: asyncFn('redis#incr'),
// 	get: asyncFn('redis#get'),
// 	hmset: asyncFn('redis#hmset'),
// 	hgetall: asyncFn('redis#hgetall'),
// 	sismember: asyncFn('redis#sismember'),
// 	sadd: asyncFn('redis#sadd'),
// 	hincrby: asyncFn('redis#hincrby'),
// 	ping: asyncFn('redis#ping'),
// 	on: asyncFn('redis#on')
// };
const getRedis = jest.fn().mockResolvedValue(redis);
jest.doMock('../services/redis/getRedis', () => getRedis);

module.exports = redis;
