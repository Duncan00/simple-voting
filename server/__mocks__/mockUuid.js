'use strict';

// make uuid predictable in snapshots
jest.mock('uuid/v4', () => {
	let count = 0;
	return () => `uuid-${count++}`;
});
