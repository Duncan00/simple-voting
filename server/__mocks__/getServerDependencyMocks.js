'use strict';

/**
 *  Setup mocks for system boundaries (third party, data storage, etc) for easier testing
 */

const mockRedis = require('./mockRedis');

require('./mockUuid');

module.exports = {
	redis: mockRedis
};
