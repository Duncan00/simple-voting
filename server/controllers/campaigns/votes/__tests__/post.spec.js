'use strict';

const supertest = require('supertest');
const _ = require('lodash');
const mockdate = require('mockdate');
const initServer = () => require('../../../../initServer')();
const getServerDependencyMocks = () =>
	require('../../../../__mocks__/getServerDependencyMocks');

describe('POST /v1/campaigns/:id/votes', () => {
	let server;

	const campaign_record = {
		id: 'fa044192-db7e-4fd3-9334-c9344360286b',
		title: 'Who is the best NBA player in the history',
		start_date: '2019-01-01',
		end_date: '2019-01-31',
		'number_of_votes:Michael Jordan': '0',
		'number_of_votes:Tim Duncan': '0'
	};

	beforeEach(async () => {
		jest.resetModules();
		mockdate.set('2019-01-01');
		const {redis} = getServerDependencyMocks();
		await redis.hmset(`campaign:{${campaign_record.id}}`, campaign_record);
		server = await initServer();
	});

	afterEach(() => {
		server.close();
		mockdate.reset();
	});

	describe('When vote with valid input and valid date', () => {
		const request_body = {
			hkid: 'A123456(3)',
			candidate: {
				name: 'Michael Jordan'
			}
		};

		test('201 - Successfully created', async () => {
			const {body, status} = await supertest(server)
				.post(`/v1/campaigns/${campaign_record.id}/votes`)
				.send(request_body);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(201);

			const {redis} = getServerDependencyMocks();
			const updated_campaign = await redis.hgetall(
				`campaign:{${campaign_record.id}}`
			);
			const voted_campaign_hkid_set = await redis.smembers(
				`voted:{${campaign_record.id}}`
			);
			const voted_candidate_hkid_set = await redis.smembers(
				`voted:{${campaign_record.id}}:${request_body.candidate.name}`
			);
			expect(updated_campaign).toMatchSnapshot('updated_campaign');
			expect(voted_campaign_hkid_set).toMatchSnapshot(
				'voted_campaign_hkid_set'
			);
			expect(voted_candidate_hkid_set).toMatchSnapshot(
				'voted_candidate_hkid_set'
			);
		});
	});

	describe('When vote with valid input and valid date concurrently', () => {
		const request_body = {
			hkid: 'A123456(3)',
			candidate: {
				name: 'Michael Jordan'
			}
		};

		test('201 - Successfully created for only 1 of the request', async () => {
			function genRequest() {
				return supertest(server)
					.post(`/v1/campaigns/${campaign_record.id}/votes`)
					.send(request_body);
			}

			const responses = await Promise.all([genRequest(), genRequest()]);

			expect(
				responses
					.map(res => ({status: res.status, body: res.body}))
					.sort()
			).toMatchSnapshot('responses');

			const {redis} = getServerDependencyMocks();
			const updated_campaign = await redis.hgetall(
				`campaign:{${campaign_record.id}}`
			);
			const voted_campaign_hkid_set = await redis.smembers(
				`voted:{${campaign_record.id}}`
			);
			const voted_candidate_hkid_set = await redis.smembers(
				`voted:{${campaign_record.id}}:${request_body.candidate.name}`
			);
			expect(updated_campaign).toMatchSnapshot('updated_campaign');
			expect(voted_campaign_hkid_set).toMatchSnapshot(
				'voted_campaign_hkid_set'
			);
			expect(voted_candidate_hkid_set).toMatchSnapshot(
				'voted_candidate_hkid_set'
			);
		});
	});

	describe('When vote with non-exist campaign_id', () => {
		const request_body = {
			hkid: 'A123456(3)',
			candidate: {
				name: 'Michael Jordan'
			}
		};

		test('404 - CAMPAIGN_NOT_FOUND', async () => {
			const non_exist_campaign_id =
				'e016e4f6-1b52-45ff-bced-79bbd8478a16';

			const {body, status} = await supertest(server)
				.post(`/v1/campaigns/${non_exist_campaign_id}/votes`)
				.send(request_body);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(404);
		});
	});

	describe('When vote with non-exist candidate name', () => {
		const request_body = {
			hkid: 'A123456(3)',
			candidate: {
				name: 'Non-exist candidate'
			}
		};

		test('404 - CANDIDATE_NOT_FOUND', async () => {
			const {body, status} = await supertest(server)
				.post(`/v1/campaigns/${campaign_record.id}/votes`)
				.send(request_body);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(404);
		});
	});

	describe('When vote with already voted HKID', () => {
		const request_body = {
			hkid: 'A123456(3)',
			candidate: {
				name: 'Michael Jordan'
			}
		};

		test('409 - ALREADY_VOTED', async () => {
			const {redis} = getServerDependencyMocks();
			// hashed HKID
			redis.sadd(
				`voted:{${campaign_record.id}}`,
				'ff4147ac6a03a3bcf124f9f49bc12e2bf3c0b2abe029a4dfaab7c39ae1f98309'
			);
			mockdate.set('2019-01-01');

			const {body, status} = await supertest(server)
				.post(`/v1/campaigns/${campaign_record.id}/votes`)
				.send(request_body);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(409);

			mockdate.reset();
		});
	});

	describe('When vote with invalid HKID', () => {
		const request_body = {
			hkid: 'A123456(A)',
			candidate: {
				name: 'Michael Jordan'
			}
		};

		test('422 - INVALID_HKID', async () => {
			const {body, status} = await supertest(server)
				.post(`/v1/campaigns/${campaign_record.id}/votes`)
				.send(request_body);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(422);
		});
	});

	describe('When vote with invalid date', () => {
		const request_body = {
			hkid: 'A123456(A)',
			candidate: {
				name: 'Michael Jordan'
			}
		};

		test('422 - INVALID_DATE', async () => {
			const {redis} = getServerDependencyMocks();
			mockdate.set('2019-12-31');

			const {body, status} = await supertest(server)
				.post(`/v1/campaigns/${campaign_record.id}/votes`)
				.send(request_body);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(422);
		});
	});
});
