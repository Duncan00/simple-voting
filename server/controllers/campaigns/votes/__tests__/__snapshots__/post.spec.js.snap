// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /v1/campaigns/:id/votes When vote with already voted HKID 409 - ALREADY_VOTED 1`] = `
Object {
  "data": Object {},
  "meta": Object {
    "code": 40902,
    "errors": Array [],
    "message": "This HKID has already voted this campaign.",
    "type": "ALREADY_VOTED",
  },
}
`;

exports[`POST /v1/campaigns/:id/votes When vote with invalid HKID 422 - INVALID_HKID 1`] = `
Object {
  "data": Object {},
  "meta": Object {
    "code": 42203,
    "errors": Array [],
    "message": "HKID is invalid.",
    "type": "HKID_INVALID",
  },
}
`;

exports[`POST /v1/campaigns/:id/votes When vote with invalid date 422 - INVALID_DATE 1`] = `
Object {
  "data": Object {},
  "meta": Object {
    "code": 42203,
    "errors": Array [],
    "message": "HKID is invalid.",
    "type": "HKID_INVALID",
  },
}
`;

exports[`POST /v1/campaigns/:id/votes When vote with non-exist campaign_id 404 - CAMPAIGN_NOT_FOUND 1`] = `
Object {
  "data": Object {},
  "meta": Object {
    "code": 40401,
    "errors": Array [],
    "message": "Campaign is not found.",
    "type": "NotFound",
  },
}
`;

exports[`POST /v1/campaigns/:id/votes When vote with non-exist candidate name 404 - CANDIDATE_NOT_FOUND 1`] = `
Object {
  "data": Object {},
  "meta": Object {
    "code": 40402,
    "errors": Array [],
    "message": "Candidate is not found.",
    "type": "NotFound",
  },
}
`;

exports[`POST /v1/campaigns/:id/votes When vote with valid input and valid date 201 - Successfully created 1`] = `
Object {
  "data": Object {
    "candidates": Array [
      Object {
        "name": "Michael Jordan",
        "number_of_votes": 1,
        "voted": false,
      },
      Object {
        "name": "Tim Duncan",
        "number_of_votes": 0,
        "voted": false,
      },
    ],
    "end_date": Array [
      "2019-01-31",
    ],
    "id": Array [
      "fa044192-db7e-4fd3-9334-c9344360286b",
    ],
    "start_date": Array [
      "2019-01-01",
    ],
    "title": Array [
      "Who is the best NBA player in the history",
    ],
  },
  "meta": Object {
    "code": 20100,
    "message": "Created",
  },
}
`;

exports[`POST /v1/campaigns/:id/votes When vote with valid input and valid date 201 - Successfully created: updated_campaign 1`] = `
Object {
  "end_date": "2019-01-31",
  "id": "fa044192-db7e-4fd3-9334-c9344360286b",
  "number_of_votes:Michael Jordan": "1",
  "number_of_votes:Tim Duncan": "0",
  "start_date": "2019-01-01",
  "title": "Who is the best NBA player in the history",
}
`;

exports[`POST /v1/campaigns/:id/votes When vote with valid input and valid date 201 - Successfully created: voted_campaign_hkid_set 1`] = `
Array [
  "ff4147ac6a03a3bcf124f9f49bc12e2bf3c0b2abe029a4dfaab7c39ae1f98309",
]
`;

exports[`POST /v1/campaigns/:id/votes When vote with valid input and valid date 201 - Successfully created: voted_candidate_hkid_set 1`] = `
Array [
  "ff4147ac6a03a3bcf124f9f49bc12e2bf3c0b2abe029a4dfaab7c39ae1f98309",
]
`;
